### ISSUE TITLE
nondeterministic timeouts with high-cpu usage greenlets

### ISSUE THREAD

**Issue by denik** [gevent/gevent#97]
Libev (and libevent) uses a cached time from the beginning of the event dispatch loop when setting timeouts.  If a greenlet does something CPU-bound followed by a connect with a timeout, the actual timeout value will be (requested_timeout - (time spent in cpu-bound greenlet)).

For example, if I have a greenlet that retrieves a bunch of things from a database, takes 200 milliseconds to process that list (of CPU-time, not blocking operations), then does a socket connect with a 100 millisecond timeout, that connect will fail immediately with a timeout.  

I can avoid this by explicitly yielding after/during cpu-bound sections of greenlet execution, but that is really ugly and unreliable.  Under high enough concurrency, this approach will fail as well.  

I'm not sure how best to fix this.  The point of the caching is to avoid doing large amounts of system calls, but without dumping the cached time, there's no way to avoid this issue.  Maybe the solution is modifying libev to allow refreshing the cached time for certain callbacks and exposing that via gevent somehow.

<b>What steps will reproduce the problem?</b>
See attached reproduction script

<b>What is the expected output? What do you see instead?</b>
timeout should either not trigger or trigger at the full 500ms.  Instead, it triggers at around 1ms.

<b>What version of the gevent are you using?</b>
Occurs in pre-1.0 versions using libevent and in post 1.0 versions using libev.

<b>What version of libevent are you using?</b>

<b>On what operating system?</b>

<b>On what Python?</b>
2.4.3 and 2.7.1

Which event loop are you running? (Paste the output of python -c &quot;from
gevent.core import *; print get_method(), get_version()&quot;)
epoll

<b>Please provide any additional information below.</b>

<p><a href="http://code.google.com/p/gevent/issues/detail?id=97">Reported by shaun@meebo-inc.com.</a></p><hr/><h2>earlier comments</h2>
<p><strong>creotiv said, at 2011-09-20T20:16:07.000Z:</strong></p>
<p>When you use "while time.time() < end: pass" you totaly block event loop, and because of this you get time problem.</p><p><strong>shaun@meebo-inc.com said, at 2011-09-20T20:48:51.000Z:</strong></p>
<p>Yup, I'm aware.  The time.time is specifically to force execution on this greenlet and not allow a yield.  The reproduction code is contrived, obviously, but the problem is real and manifests is actual, non-contrived production cases.

The issue we're seeing is while processing a large set of responses from a database query, followed by making another database query, which instantly times out.

The repro script was simplest way I could recreate it.  While you'd never do the while time.time(): business is real code, the problem manifests with any largely cpu-bound greenlet coupled with timeouts. </p><p><strong>srlindsay said, at 2011-09-21T01:42:28.000Z:</strong></p>
<p>I'm going to go ahead and patch libev to not use the cached time when creating timeouts, probably protected by a compile time flag or something.  It doesn't seem like the ideal solution, but it will at least be a decent short-term fix at the expense of more system calls.  I'll post the patch when I have something working.</p><p><strong>shaun@meebo-inc.com said, at 2011-09-23T01:13:43.000Z:</strong></p>
<p>I patched libev to not use the cached time in ev_timer_start.  This isn't a good long-term solution, but fixes the timeout issue at the cost of performance (due to requiring a syscall per timer).

A better long-term solution to the issue will be modifying libev itself to defer timer starts to the end of the event dispatch cycle.  Refreshing the cached time then running ev_timer_start on all the deferred timers will make the actual timeout value independent of the time taken in event dispatch.

This issue is more a problem with libev itself, so I'll bring it up with the libev developers.</p><p><strong>Denis.Bilenko said, at 2011-10-07T09:09:11.000Z:</strong></p>
<p>Here's the mailing list thread related to this: http://groups.google.com/group/gevent/t/77c67e6b2bdeaed7 (I've replied it here)</p><p><strong>shaun@meebo-inc.com said, at 2011-10-11T23:28:17.000Z:</strong></p>
<p>Here's a patch for libev to defer ev_timer_starts until the end of the event dispatch cycle, then start them all in a batch.  Seems okay thus far in my tests.  I'll also submit the patch to the libev folks as well.  I'll also reply on the mailing list thread.</p><p><strong>Denis.Bilenko said, at 2011-12-20T17:06:43.000Z:</strong></p>
<p>I've committed a fix that just calls ev_now_update() before starting timers: https://bitbucket.org/denis/gevent/changeset/71e0ace5cfa3

Please try it out.

One potential downside is that starting a timer is a bit slower now: my rough benchmarks show that creating and starting a timer time increased from 1.5ns to 2-2.5ns.

However, it's probably won't make any difference and if it does, then it's possible to disable this behaviour.

</p>
